[NNC] Fix an issue in Cuda fusion with fp16 scalar vars coerced to float

nickgg opened the pull request:
Fixes an issue where fp16 scalars created by the registerizer could be referenced as floats - causing invalid conversions which would crash in the NVRTX compile. I also noticed that we were inserting patterns like `float(half(float(X)))` and added a pass to collapse those down inside the CudaHalfScalarRewriter.

Fixes #47138

nickgg committed:
[NNC] Fix an issue with half-scalar vars coerced to float

eellison commented on torch/csrc/jit/tensorexpr/cuda_half_support.h:
```diff
@@ -85,12 +85,32 @@ class CudaHalfScalarRewriter : public IRMutator {
     auto it = var_map.find(v);
     if (it != var_map.end()) {
-      return it->second;
+      return new Cast(kHalf, it->second);
```
Mind explaining why we need to cast the original var back to half, after we've already gone through the work of casting it from half to float ?

nickgg replied:
Sure, the case in the issue above is something like this:
```
A[x] = B[x];
... more stuff using B[x]
```

Which after half replacement is like:

```
A[x] = half(float(B[x]));
... more stuff using float(B[x]);
```

Then registerization - which doesn't know that Half is not allowed:

```
half B_1 = B[x];
A[x] = B_1;
... more stuff using float(B_1)
```

There are again Halfs, so we have the CudaHalfScalarRewriter:

```
float B_1 = float(B[x]);
A[x] = B_1;
... more stuff using B_1
```

And this is where it crashes, because A[x] is Half and B_1 is Float. By wrapping B_1 with Half() we cover this case where there isn't an enclosing cast. If it's not a direct write to Half, then it would already be wrapped by a Float cast from the CudaHalfChecker.

As I write this I'm wondering why it's not possible to delay Half checking until registerization which should solve this (and other) weirdnesses. I remember trying but can't remember why it didn't work. Let me have a look.

nickgg replied:
Oh, it was because we need to insert the half type definitions before the function begins... Still maybe easier to move all the logic in to the rewriter.

eellison commented on test/test_tensorexpr.py:
```diff
@@ -1262,14 +1264,52 @@ def bias_gelu(bias, y):
+                for param in net_ref.parameters():
+                    param.grad = None
+
+                net(inp).mean().backward()
```
imo it would be nice to rewrite this without calling backward, since autograd is complex & makes this test a little harder to reason abt

nickgg replied:
I think original bug may have required it. The issue didn't appear consistently in the NNC output (which is deterministic) so it's something above this layer.

Krovatkin commented on torch/csrc/jit/tensorexpr/cuda_half_support.h:
```diff
@@ -85,12 +85,32 @@ class CudaHalfScalarRewriter : public IRMutator {
     return v;
   }
 
+  const Expr* mutate(const Cast* v) override {
```
do we typically keep this kind of peepholes in individual passes or move them to simplifier?

nickgg replied:
Normally they'd be in the simplifier, but I'm not confident we always want to eliminate Float(Half(X)) since it does have an affect (reducing precision). What do you think.

Krovatkin reviewed (approved):
:shipit:

nickgg commented on torch/csrc/jit/tensorexpr/cuda_half_support.h:
```diff
@@ -85,12 +85,32 @@ class CudaHalfScalarRewriter : public IRMutator {
+
+    const Cast* cast_child = dynamic_cast<const Cast*>(child);
+    if (cast_child) {
+      if ((v->dtype().scalar_type() == ScalarType::Float ||
```
Ah, not all casts are safe to eliminate - e.g `Float(Int(X))` is not safe to replace with `X`. Rather than work out what the correct behavior in all cases I decided to keep this limited to float and half since we're casting all halfs in this backend anyway.

eellison replied:
Hm, i may have misread this I read this as int(float(int_val)) - i think any cast from a type1 -> type2  -> type1 can always be removed right ?

nickgg replied:
I don't think so, `Float(Int(X))` isn't safe when X is Float, neither is `Int(Float(Int(X)))` when X is a floating type.

nickgg commented:
Okay, I put up a new version which moves all the modifications to the HalfRewriter and out of the HalfChecker (which is now an IRVisitor). This allowed some minor simplifications (such as the extra Half cast @eellison pointed out).

Also since you each mentioned it was clumsy I changed the Cast elimination code to use `is_floating_point` rather than explicitly looking for Float and Half. Does this help.

Also I fixed the issue with vectorizing since I was refactoring anyway.

eellison commented on torch/csrc/jit/tensorexpr/cuda_half_support.h:
```diff
@@ -50,8 +74,16 @@ class CudaHalfChecker : public IRMutator {
     if (v->dtype().scalar_type() == ScalarType::Half) {
       if (inserted_half_casts_.count(v) < 1) {
-        // TODO: discards lanes.
-        return new Cast(kFloat, child);
```
Reason for this change ? Retaining cast to float seemed more semantics-preserving

eellison commented on torch/csrc/jit/tensorexpr/cuda_half_support.h:
```diff
@@ -50,8 +74,16 @@ class CudaHalfChecker : public IRMutator {
+    // Remove Half(Float()) and friends.
+    const Cast* cast_child = dynamic_cast<const Cast*>(child);
+    if (cast_child) {
+      if (v->dtype().is_floating_point() &&
```
I agree with your earlier comment 
> Normally they'd be in the simplifier, but I'm not confident we always want to eliminate Float(Half(X)) since it does have an affect (reducing precision).

What about specifically the case of casting from a type to a new type and back again ? I think that can generally be removed without reducing precision.

nickgg replied:
I don't see how, consider: `Int(Char(Int(X)))` when X is FP. Both the cast to Int and the cast to Char have effects here.

I'm not opposed to the idea of a general casting simplification but I think it's somewhat complex and don't want to roll it into this fix.

eellison replied:
Ah true, i guess it's just lower precision -> higher precision -> lower precision where both types are floating point or both are integers

eellison reviewed (commented):
lgtm! thx for quick fix & explanations! a couple small comments. maybe worth getting a review from someone a little more familiar with NNC 🤷

zou3519 commented:
This broke some windows tests:
- https://app.circleci.com/pipelines/github/pytorch/pytorch/235345/workflows/93896a08-58a7-438d-9a22-73109aae8644/jobs/8709371
- https://app.circleci.com/jobs/github/pytorch/pytorch/8709399

So I'm reverting it
